### ISSUE TITLE
The parameter entrypoints.<name>.http.tls.domains is not used if rules in routers are only PathPrefix

### ISSUE THREAD

**Issue by timurb** [traefik/traefik#7438]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

- When I specify TLS domains for Letsencrypt in entrypoints section
- and I have no domains specified for Letsencrypt in router section
- and all of my router rules are based on `PathPrefix`
- then the certificate is not requested from Letsencrypt

### What did you expect to see?

I would expect domains for cert generation to be picked up from entrypoints section

### My docker-compose file

```
version: '3'

services:
  reverse-proxy:
    # The official v2 Traefik docker image
    image: traefik:v2.3

    restart: always

    # Don't put comments in this block as it is a single long string
    command:
      - "--log.level=DEBUG"
      - "--providers.docker"

      - "--entryPoints.websecure.address=:443"
      - "--entrypoints.websecure.http.tls=true"
      - "--entrypoints.websecure.http.tls.certResolver=letsencrypt"
      - "--entrypoints.websecure.http.tls.domains[0].main=domain.example.com"   # put your domain name here

      - "--certificatesresolvers.letsencrypt.acme.email=your@emailcom"
      - "--certificatesresolvers.letsencrypt.acme.storage=/traefik/acme.json"
      - "--certificatesresolvers.letsencrypt.acme.caserver=https://acme-staging-v02.api.letsencrypt.org/directory"
      - "--certificatesresolvers.letsencrypt.acme.tlschallenge=true"

    ports:
      - "443:443"

    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro  # So that Traefik can listen to the Docker events
      - /traefik:/traefik

    labels:
      - "traefik.http.routers.dashboard.entrypoints=dashboard"
      - "traefik.http.routers.dashboard.tls=true"
      - "traefik.http.routers.dashboard.tls.certresolver=letsencrypt"
      - "traefik.http.routers.dashboard.rule=(PathPrefix(`/api`) || PathPrefix(`/dashboard`))"

  httpbin:
    # A container that exposes an API to show its IP address
    image: kennethreitz/httpbin

    restart: always

    labels:
      - "traefik.http.routers.app.rule=PathPrefix(`/`)"
      - "traefik.http.routers.app.entrypoints=websecure"
      - "traefik.http.routers.app.tls.certresolver=letsencrypt"
#      - "traefik.http.routers.app.tls.domains[0].main=domain.example.com"       # put your domain name here
      - "traefik.http.services.app.loadbalancer.server.port=80"
```

With this config certificates are not generated. 
To trigger certificate generation do either of these:
* comment out any of the lines containing `PathPrefix`
* uncomment the line `traefik.http.routers.app.tls.domains[0].main`

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
$ docker-compose up
Creating network "ubuntu_default" with the default driver
Creating ubuntu_reverse-proxy_1 ... done
Creating ubuntu_httpbin_1       ... done
Attaching to ubuntu_httpbin_1, ubuntu_reverse-proxy_1
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Configuration loaded from flags."
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Traefik version 2.3.1 built on 2020-09-29T15:49:06Z"
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=debug msg="Static configuration loaded {\"global\":{\"checkNewVersion\":true},\"serversTransport\":{\"maxIdleConnsPerHost\":200},\"entryPoints\":{\"websecure\":{\"address\":\":443\",\"transport\":{\"lifeCycle\":{\"graceTimeOut\":10000000000},\"respondingTimeouts\":{\"idleTimeout\":180000000000}},\"forwardedHeaders\":{},\"http\":{\"tls\":{\"certResolver\":\"letsencrypt\",\"domains\":[{\"main\":\"mayak.timurb.ru\"}]}}}},\"providers\":{\"providersThrottleDuration\":2000000000,\"docker\":{\"watch\":true,\"endpoint\":\"unix:///var/run/docker.sock\",\"defaultRule\":\"Host(`{{ normalize .Name }}`)\",\"exposedByDefault\":true,\"swarmModeRefreshSeconds\":15000000000}},\"log\":{\"level\":\"DEBUG\",\"format\":\"common\"},\"certificatesResolvers\":{\"letsencrypt\":{\"acme\":{\"email\":\"erthad@gmail.com\",\"caServer\":\"https://acme-staging-v02.api.letsencrypt.org/directory\",\"storage\":\"/traefik/acme.json\",\"keyType\":\"RSA4096\",\"tlsChallenge\":{}}}}}"
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="\nStats collection is disabled.\nHelp us improve Traefik by turning this feature on :)\nMore details on: https://doc.traefik.io/traefik/contributing/data-collection/\n"
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Starting provider aggregator.ProviderAggregator {}"
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=debug msg="Start TCP Server" entryPointName=websecure
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Starting provider *docker.Provider {\"watch\":true,\"endpoint\":\"unix:///var/run/docker.sock\",\"defaultRule\":\"Host(`{{ normalize .Name }}`)\",\"exposedByDefault\":true,\"swarmModeRefreshSeconds\":15000000000}"
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Starting provider *acme.Provider {\"email\":\"erthad@gmail.com\",\"caServer\":\"https://acme-staging-v02.api.letsencrypt.org/directory\",\"storage\":\"/traefik/acme.json\",\"keyType\":\"RSA4096\",\"tlsChallenge\":{},\"ResolverName\":\"letsencrypt\",\"store\":{},\"ChallengeStore\":{}}"
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Testing certificate renew..." providerName=letsencrypt.acme
reverse-proxy_1  | time="2020-10-20T11:21:18Z" level=info msg="Starting provider *traefik.Provider {}"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Configuration received from provider letsencrypt.acme: {\"http\":{},\"tls\":{}}" providerName=letsencrypt.acme
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Configuration received from provider internal: {\"http\":{\"services\":{\"noop\":{}},\"models\":{\"websecure\":{\"tls\":{\"certResolver\":\"letsencrypt\",\"domains\":[{\"main\":\"mayak.timurb.ru\"}]}}}},\"tcp\":{},\"tls\":{}}" providerName=internal
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No default certificate, generating one"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Provider connection established with docker 19.03.8 (API 1.40)" providerName=docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Configuration received from provider docker: {\"http\":{\"routers\":{\"app\":{\"entryPoints\":[\"websecure\"],\"service\":\"app\",\"rule\":\"PathPrefix(`/`)\",\"tls\":{\"certResolver\":\"letsencrypt\"}},\"dashboard\":{\"entryPoints\":[\"dashboard\"],\"service\":\"reverse-proxy-ubuntu\",\"rule\":\"(PathPrefix(`/api`) || PathPrefix(`/dashboard`))\",\"tls\":{\"certResolver\":\"letsencrypt\"}}},\"services\":{\"app\":{\"loadBalancer\":{\"servers\":[{\"url\":\"http://172.31.0.2:80\"}],\"passHostHeader\":true}},\"reverse-proxy-ubuntu\":{\"loadBalancer\":{\"servers\":[{\"url\":\"http://172.31.0.3:80\"}],\"passHostHeader\":true}}}},\"tcp\":{},\"udp\":{}}" providerName=docker
httpbin_1        | [2020-10-20 11:21:19 +0000] [1] [INFO] Starting gunicorn 19.9.0
httpbin_1        | [2020-10-20 11:21:19 +0000] [1] [INFO] Listening at: http://0.0.0.0:80 (1)
httpbin_1        | [2020-10-20 11:21:19 +0000] [1] [INFO] Using worker: gevent
httpbin_1        | [2020-10-20 11:21:19 +0000] [8] [INFO] Booting worker with pid: 8
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="entryPoint \"dashboard\" doesn't exist" routerName=dashboard@docker entryPointName=dashboard
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="no valid entryPoint for this router" routerName=dashboard@docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating middleware" middlewareName=pipelining middlewareType=Pipelining entryPointName=websecure routerName=app@docker serviceName=app
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating load-balancer" serviceName=app entryPointName=websecure routerName=app@docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating server 0 http://172.31.0.2:80" entryPointName=websecure routerName=app@docker serviceName=app serverName=0
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Added outgoing tracing middleware app" middlewareType=TracingForwarder entryPointName=websecure routerName=app@docker middlewareName=tracing
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating middleware" middlewareName=traefik-internal-recovery middlewareType=Recovery entryPointName=websecure
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="entryPoint \"dashboard\" doesn't exist" routerName=dashboard@docker entryPointName=dashboard
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="no valid entryPoint for this router" routerName=dashboard@docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No default certificate, generating one"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="entryPoint \"dashboard\" doesn't exist" routerName=dashboard@docker entryPointName=dashboard
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="no valid entryPoint for this router" routerName=dashboard@docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating middleware" entryPointName=websecure serviceName=app routerName=app@docker middlewareName=pipelining middlewareType=Pipelining
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating load-balancer" serviceName=app routerName=app@docker entryPointName=websecure
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating server 0 http://172.31.0.2:80" serverName=0 routerName=app@docker entryPointName=websecure serviceName=app
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Added outgoing tracing middleware app" middlewareType=TracingForwarder entryPointName=websecure routerName=app@docker middlewareName=tracing
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="Creating middleware" entryPointName=websecure middlewareName=traefik-internal-recovery middlewareType=Recovery
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="entryPoint \"dashboard\" doesn't exist" entryPointName=dashboard routerName=dashboard@docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=error msg="no valid entryPoint for this router" routerName=dashboard@docker
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No default certificate, generating one"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No domain parsed in provider ACME" providerName=letsencrypt.acme routerName=app@docker rule="PathPrefix(`/`)"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No domain parsed in provider ACME" providerName=letsencrypt.acme routerName=dashboard@docker rule="(PathPrefix(`/api`) || PathPrefix(`/dashboard`))"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No domain parsed in provider ACME" providerName=letsencrypt.acme routerName=app@docker rule="PathPrefix(`/`)"
reverse-proxy_1  | time="2020-10-20T11:21:19Z" level=debug msg="No domain parsed in provider ACME" providerName=letsencrypt.acme rule="(PathPrefix(`/api`) || PathPrefix(`/dashboard`))" routerName=dashboard@docker
```
For successful certificate generation please comment/uncomment the appropriate loglines in the docker-compose.yaml and see it yourself.

**Comment by timurb** [comment]
Actually if I don't uncomment the line `"traefik.http.routers.app.tls.domains[0].main=domain.example.com"` the certificate is not requested as well (which is contrary to documentation here https://doc.traefik.io/traefik/reference/static-configuration/cli/) — the private keys is created but that's all.

With all router rules having `PathPrefix` even private key is not created and `acme.json` is an empty json file.

**Comment by jbdoumenjou** [comment]
Hi @timurb ,

I suppose the issue comes from the configuration.
As said in the [documentation](https://doc.traefik.io/traefik/routing/entrypoints/#tls): 

> If a TLS section (i.e. any of its fields) is user-defined, then the default configuration does not apply at all.

So, as you defined `- "traefik.http.routers.app.tls.certresolver=letsencrypt"`, the default configuration will not be applied.
Try without any label with tls configuration or with all labels needed.

Could you please confirm this hypothesis ?

**Comment by jbdoumenjou** [comment]
Thanks for confirming the issue. :+1: 

So I close the issue accordingly.

If you want clues on your configuration, you can use [api](https://doc.traefik.io/traefik/operations/api/#endpoints).
If you would like to discuss this further, please visit the [community forum](https://community.traefik.io/).

**Comment by timurb** [comment]
I don't think the issue is resolved and I'm sorry if I didn't do well with the new issue description.

The issue should be: produce debug/info message in logs if the configuration override is happening.
And probably one more: produce warning message in logs if incomplete TLS configuration is found for the endpoint.

Do you think we should reopen this issue to keep the context or create new one(s) for visibility and detailed discussion?

**Comment by timurb** [comment]
Actually I see the message `No domain parsed in provider ACME` but I'm able to understand its meaning only after you explained it so there still needs to be some clue in log messages that something is missing in the configuration:
- producing a new logline that the override has happened for this config and parent configuration was ignored — as a standalone message
- extending the current error message to something like `No domain parsed in provider ACME. Did you define "domains" section?` and/or probably raising the severity to warning/error.
- producing warning message if TLS configuration for the endpoint is incomplete
- mark the documentation snippet you quoted with a large /!\ or (i) symbol because a small note in the middle of a paragraph effectively nullifies it completely in certain circumstances  

My idea and goal behind that is when the next person faces the same issue they are able to understand it without creating the ticket/forum thread etc. Currently you have to see (and have at hand) a couple of successful TLS installations to be able to understand what is happening in this scenario without external support.

**Comment by jbdoumenjou** [comment]
Hi @timurb,

Thanks for the feedback.

The logging part is not easy because we can't know when the configuration is incomplete or if it is attended.
And if we choose to log any potential misconfiguration, it will be a spam mechanism :wink: 

It would be great if you can open a specific issue for the documentation and then work on PR that will fix the documentation.

Then, if you think we can do more, you can create a more open issue on other improvements that needs more thought.

WDYT ?
